### PR TITLE
resource_arm_application_gateway: add exclusion configuration for waf…

### PR DESCRIPTION
**Pull request by bs-matil**
This PR adds an `exclusion` block to the `waf_configuration` this enables the user to set exclusions from WAF scanning.

### CODE REVIEW

File: azurerm/resource_arm_application_gateway.go [hunk 1]
```diff
@@ -2620,6 +2660,7 @@ func expandApplicationGatewayWafConfig(d *schema.ResourceData) *network.Applicat
 		FileUploadLimitInMb:    utils.Int32(int32(fileUploadLimitInMb)),
 		RequestBodyCheck:       utils.Bool(requestBodyCheck),
 		MaxRequestBodySizeInKb: utils.Int32(int32(maxRequestBodySizeInKb)),
+		Exclusions:             &exclusions,
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway.go]: We want to have the above code in a separate expand function, similar to what I suggested in the other review

File: azurerm/resource_arm_application_gateway.go [hunk 2]
```diff
@@ -2657,6 +2698,29 @@ func flattenApplicationGatewayWafConfig(input *network.ApplicationGatewayWebAppl
...
+			}
+
+			if exclusionList.Selector != nil {
+				exclusionListOutput["selector"] = *exclusionList.Selector
+			}
+
+			exclusionLists = append(exclusionLists, exclusionListOutput)
+		}
+
+		output["exclusion"] = exclusionLists
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway.go]: We want to have the above code in a flatten method, similar to what I suggested in the other review

File: azurerm/resource_arm_application_gateway_test.go [hunk 3]
```diff
@@ -536,6 +536,50 @@ func TestAccAzureRMApplicationGateway_connectionDraining(t *testing.T) {
...
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.2.selector_match_operator", "StartsWith"),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.2.selector", "ORIGIN"),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.3.match_variable", "RequestHeaderNames"),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.3.selector_match_operator", "Contains"),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.3.selector", "ORIGIN"),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.4.match_variable", "RequestHeaderNames"),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.4.selector_match_operator", ""),
+					resource.TestCheckResourceAttr(resourceName, "waf_configuration.0.exclusion.4.selector", ""),
+				),
+			},
```
Comment (mbfrahry) [inline azurerm/resource_arm_application_gateway_test.go]: Let's add another test step here to confirm that we can update the added fields properly

File: website/docs/r/application_gateway.html.markdown [hunk 4]
```diff
@@ -397,6 +397,18 @@ A `waf_configuration` block supports the following:
 
 * `max_request_body_size_kb` - (Optional) The Maximum Request Body Size in KB.  Accepted values are in the range `1`KB to `128`KB.  Defaults to `128`KB.
 
+* `exclusion` - (Optional) one or many `exclusion` blocks as defined below.
```
Comment (mbfrahry) [inline website/docs/r/application_gateway.html.markdown]:
```suggestion
* `exclusion` - (Optional) One or more `exclusion` blocks as defined below.
```

File: azurerm/resource_arm_application_gateway.go [hunk 5]
```diff
@@ -919,6 +919,29 @@ func resourceArmApplicationGateway() *schema.Resource {
...
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"match_variable": {
+										Type:     schema.TypeString,
+										Required: true,
+									},
+
+									"selector_match_operator": {
+										Type:     schema.TypeString,
+										Optional: true,
```
[the remark this answers is not in the document]
Comment (bs-matil) [reply azurerm/resource_arm_application_gateway.go]: actually the problem is that it not "really" defaults to equals any. The reason is that the API is diffrent than the azure ui. In the UI you would set "equals any" in the API an nil feaule is treated as equals_any. Actually I don't know how to face this best to the user.

**Comment by bs-matil** [comment]
@mbfrahry I submitted fixes for "all" issues. Please review again.

File: azurerm/resource_arm_application_gateway.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bs-matil) [reply azurerm/resource_arm_application_gateway.go]: I decided to change the documentation and remove the reference to the ui at all

**Comment by mbfrahry** [comment]
Thanks @bs-matil! The conflicts just need to be resolved from between this and your other PR

**Comment by bs-matil** [comment]
@mbfrahry  closed in favor of #3407

**Pull request closed without merging**